knowrob_common/prolog/owl.pl

```diff
@@ -709,6 +788,79 @@
 owl_has_direct(S, P, O) :-
 	owl_use_rule(S, P, O).
 
+%% Simplest branch: find an explicitly stored rdf triple (S, P, O)
+owl_has_direct_internal(S, P, O) :-
+	rdf_has(S, P, O).
+
+%% If P is bound to an object property, see if any of its PropertyChain axioms is able to produce explicitly known triples.
+%% ASSUMPTION: no circular PropertyChain axioms (example, P defined as A o B and A defined as P o B)
+owl_has_direct_internal(S, P, O) :-
+        nonvar(P),
+        owl_individual_of(P, owl:'ObjectProperty'),
+        rdf_has(P, owl:propertyChainAxiom, RDFList),
+        kr_rdf_list(RDFList, PropChain),
+% We need this because the PropChain contains a list of either property or Description individuals (where a Description individual refers to an inverse property),
+% and also we need a simple list of the props appearing in a property chain.
+        kr_get_props(PropChain, Props),
+% Ensure we avoid circular property chains
+        \+ kr_circular_chain(Props, [P]),
+        owl_has_property_chain(S, PropChain, O).
+
+%% No longer needed. We now attempt direct and inverse properties using the owl_has_direct above, and defined a predicate
+%% owl_has_direct_internal that never looks for inverse properties, to avoid infinite backtracking. This fixes the TBD below.
+% owl_has_direct(S, P, O) :-
+%	(   rdf_has(P, owl:inverseOf, P2)
+%	->  true
+%	;   rdf_has(P2, owl:inverseOf, P)
+%	),
+%	rdf_has(O, P2, S).		% TBD: must call owl_has_direct/3
+
+%% Copied above
+% owl_has_direct(S, P, O) :-
+% 	rdfs_individual_of(P, owl:'SymmetricProperty'),
+% 	rdf(O, P, S).
+
+owl_has_direct_internal(S, P, O) :-
+	owl_use_has_value(S, P, O).
+
+owl_has_property_chain(S, PropChain, O) :-
+        nonvar(S),
+        owl_has_property_chain_S2O(S, PropChain, O).
+
+owl_has_property_chain(S, PropChain, O) :-
+        var(S),
+        nonvar(O),
+        reverse(PropChain, PropChainRev),
+        owl_has_property_chain_O2S(O, PropChainRev, S).
+
+owl_has_property_chain_S2O(O, [], O).
+
+owl_has_property_chain_S2O(S, PropChain, O) :-
+        nth0(0, PropChain, P, RestChain),
+        owl_individual_of(P, owl:'ObjectProperty'),
+        owl_has(S, P, Oi),
+        owl_has_property_chain_S2O(Oi, RestChain, O).
+
+owl_has_property_chain_S2O(S, PropChain, O) :-
+        nth0(0, PropChain, P, RestChain),
+        owl_has(P, owl:inverseOf, PI),
+        owl_has(Oi, PI, S),
+        owl_has_property_chain_S2O(Oi, RestChain, O).
+
+owl_has_property_chain_O2S(S, [], S).
+
+owl_has_property_chain_O2S(O, PropChainRev, S) :-
```

Comment:
Maybe, but I'd rather not risk it. the *_S2O and *_O2S are written with the assumption (checked with var/nonvar predicates) that one of the S, O variables is bound, and the functions will start verifying the property chain from the end that is bound. So for example, if S is bound, then *_S2O is called, and it will check the chain starting from S. While it may be possible to start verifying the chain from the unbound end, I imagine it would be very inefficient as many branches will be explored before one is found that completes the chain to the bound end.